include/libtorrent/config.hpp

```diff
@@ -98,6 +98,9 @@ POSSIBILITY OF SUCH DAMAGE.
 # define TORRENT_DEPRECATED __declspec(deprecated)
 #endif
 
+// auto and decltype(auto) return types supports since MSVS2015
+// https://msdn.microsoft.com/en-us/library/hh567368.aspx
+#define TORRENT_AUTO_RETURN_TYPES 1
```

Comment:
please mention why we need this too. i.e. msvc-15 doesn't support the C++11 syntax.
Also, ideally this would only get enabled for that version of msvc